clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -0,0 +1,212 @@
+ * consumer group protocol.
+ * <p/>
+ * This keeps membership state and assignment updated in-memory, based on the heartbeat responses
+ * the member receives. It is also responsible for computing assignment for the group based on
+ * the metadata, if the member has been selected by the broker to do so.
+ */
+public class MembershipManagerImpl implements MembershipManager {
+
+    private final String groupId;
+    private final Optional<String> groupInstanceId;
+    private String memberId;
+    private int memberEpoch;
+    private MemberState state;
+    private AssignorSelection assignorSelection;
+
+    /**
+     * Assignment that the member received from the server and successfully processed
+     */
+    private ConsumerGroupHeartbeatResponseData.Assignment currentAssignment;
+
+    /**
+     * List of assignments that the member received from the server but hasn't processed yet
+     */
+    private final List<ConsumerGroupHeartbeatResponseData.Assignment> targetAssignments;
+
+    public MembershipManagerImpl(String groupId) {
+        this(groupId, null, null);
+    }
+
+    public MembershipManagerImpl(String groupId, String groupInstanceId, AssignorSelection assignorSelection) {
+        this.groupId = groupId;
+        this.state = MemberState.UNJOINED;
+        if (assignorSelection == null) {
+            setAssignorSelection(AssignorSelection.defaultAssignor());
+        } else {
+            setAssignorSelection(assignorSelection);
+        }
+        this.targetAssignments = new ArrayList<>();
+        this.groupInstanceId = Optional.ofNullable(groupInstanceId);
+    }
+
+    /**
+     * Update assignor selection for the member.
+     *
+     * @param assignorSelection New assignor selection
+     * @throws IllegalArgumentException If the provided assignor selection is null
+     */
+    public void setAssignorSelection(AssignorSelection assignorSelection) {
+        if (assignorSelection == null) {
+            throw new IllegalArgumentException("Assignor selection cannot be null");
+        }
+        this.assignorSelection = assignorSelection;
+    }
+
+    private void transitionTo(MemberState nextState) {
+        if (!this.state.equals(nextState) && !nextState.getPreviousValidStates().contains(state)) {
+            // TODO: handle invalid state transition
+            throw new RuntimeException(String.format("Invalid state transition from %s to %s",
+                    state, nextState));
+        }
+        this.state = nextState;
+    }
+
+    @Override
+    public String groupId() {
+        return groupId;
+    }
+
+    @Override
+    public Optional<String> groupInstanceId() {
+        return groupInstanceId;
+    }
+
+    @Override
+    public String memberId() {
+        return memberId;
+    }
+
+    @Override
+    public int memberEpoch() {
+        return memberEpoch;
+    }
+
+    @Override
+    public void updateState(ConsumerGroupHeartbeatResponseData response) {
```

Comment:
WDYT about a more descriptive naming: like onHeartbeatResponse().  FWIW: in the heartbeat request manager, I also need to parse the error and response in order to update the local request state. (For exponential backoff). So would it make sense to break it into two functions like: onSuccessfulHeartbeatResponse and onErrorHeartbeatResponse.  Let me know what do you think.

Reply:
in the request manager we would probably do:
{
   if (response.errorCode() == NONE) { 
 .. update some local state
 memberstate.onResponse(response) 
} else {
memberstate.onError(response.errorcode); 
}

Reply:
Uhm actually seems that the heartbeat code you have will end up being simpler if we keep it like we have it now, with a single call to a self contained `updateState` that knows how to update the state based on a HB response. It would only require a single call to `memberState.updateState(response);` and that if/else above would be simpler. WDYT? 
Totally ok with renaming to include onHeartbeatResponse as you suggested.

Reply:
the heartbeat code will also need to handle the error code, for example upon receiving NOT_COORDINATOR error, it will need to invalidate the current coordinator node, so it will need to decode the response.  It is fine for heartbeat to pass the response to the response, but i'm just nit picking about doing the same thing twice.